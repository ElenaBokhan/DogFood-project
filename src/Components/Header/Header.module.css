.header {
    display: flex;
    padding: 12px 0;
    align-items: center;
    gap: 80px;
    height: 80px;
    justify-content: space-between;
}

.profiles-icons {
    display: flex;
    width: 140px;
    justify-content: space-between;
}

.profilesIcon {
    position: relative;
}

.babl {
    position: absolute;
    right: -4px;
    top: -2px;
    padding: 2px 5px;
    background-color: #01A54E;
    color: white;
    border-radius: 50%;
    font-size: 9px;
}