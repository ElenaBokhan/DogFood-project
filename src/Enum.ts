export enum ETestId {
    ADD_PRODUCT_NAME_INPUT = 'addProductNameInput',
    ADD_PRODUCT_PRICE_INPUT = 'addProductPriceInput',
    ADD_PRODUCT_DESCRIPTION_INPUT = 'addProductDescriptionInput',
    ADD_PRODUCT_DISCOUNT_INPUT = 'addProductDiscountInput',
    ADD_PRODUCT_STOCK_INPUT = 'addProductStockInput',
    ADD_PRODUCT_WIGHT_INPUT = 'addProductWightInput',
    ADD_PRODUCT_PICTURES_INPUT = 'addProductPicturesInput',
    ADD_PRODUCT_SUBMIT_BUTTON = 'addProductSubmitButton',

    ADD_REVIEW_SUBMIT_BUTTON = 'addReviewSubmitButton',
    ADD_REVIEW_TEXT_INPUT = 'addReviewTextInput',

    ADD_TO_FAVOURITES = 'addToFavouritesButton',
    REMOVE_FROM_FAVOURITES = 'removeFromFavourites',

    BREADCRUMBS_BUTTON = 'breadcrumbsButton',

    FOOTER_MAIN_LOGO = 'footerMainLogo',
    FOOTER_MENU = 'footerMenu',
    FOOTER_CONTACTS = 'footerContacts',

    HEADER_MAIN_LOGO = 'headerMainLogo',
    HEADER_ADD_NEW_ICON = 'headerAddNewProduct',
    HEADER_FAVOURITES_ICON = 'headerFavouritesIcon',
    HEADER_CART_ICON = 'headerCartIcon',
    HEADER_PROFILE_ICON = 'headerProfileIcon',
    HEADER_LOG_OUT_ICON = 'headerLogOutIcon',
    HEADER_SEARCH_INPUT = 'headerSearchInput',

    SIGN_FORM_EMAIL_INPUT = 'signFormEmailInput',
    SIGN_FORM_PASSWORD_INPUT = 'signFormPasswordlInput',
    SIGN_FORM_SUBMIT_BUTTON = 'signFormSubmitButton',

    TITLE_PAGE = 'titlePage',
    NOT_FOUND = 'notFound',

    PRODUCT_LIKE_BUTTON = 'likeButton',
    PRODUCT_TRASH_BUTTON = 'trashButton',
    PRODUCT_NAME = 'productName',
    PRODUCT_IMAGE = 'productImage',
}
