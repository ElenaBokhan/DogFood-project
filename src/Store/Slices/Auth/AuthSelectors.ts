import {TRootState} from 'Store/configureStore';

export const accessTokenSelector = (state: TRootState) => state.auth.accessToken;
export const refreshTokenSelector = (state: TRootState) => state.auth.refreshToken;
