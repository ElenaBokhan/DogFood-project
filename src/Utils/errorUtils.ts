import {objectHasProperty} from 'Utils/utils';

export const getMessageFromError = (error: unknown, defaultErrorMessage: string) => {
    if (
        objectHasProperty(error, 'data') &&
        objectHasProperty(error.data, 'message') &&
        typeof error.data.message === 'string'
    )
        return error.data.message;
    return defaultErrorMessage;
};
