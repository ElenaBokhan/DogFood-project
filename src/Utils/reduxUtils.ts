import type {AnyAction, AsyncThunk} from '@reduxjs/toolkit';

type GenericAsyncThunk = AsyncThunk<unknown, unknown, any>;
type PendingAction = ReturnType<GenericAsyncThunk['pending']>;
type RejectedAction = ReturnType<GenericAsyncThunk['rejected']>;
type FulfilledAction = ReturnType<GenericAsyncThunk['fulfilled']>;

const hasPrefix = (action: AnyAction, prefix: string): boolean => action.type.startsWith(prefix);
const isPending = (action: PendingAction): boolean => action.type.endsWith('/pending');
const isFulfilled = (action: FulfilledAction): boolean => action.type.endsWith('/fulfilled');
const isRejected = (action: RejectedAction): boolean => action.type.endsWith('/rejected');

export const isActionPending = (prefix: string) => (action: PendingAction) => {
    return hasPrefix(action, prefix) && isPending(action);
};

export const isActionRejected = (prefix: string) => (action: RejectedAction) => {
    return hasPrefix(action, prefix) && isRejected(action);
};

export const isActionFulfilled = (prefix: string) => (action: FulfilledAction) => {
    return hasPrefix(action, prefix) && isFulfilled(action);
};
