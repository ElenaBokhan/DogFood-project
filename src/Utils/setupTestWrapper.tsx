import {router} from 'App';
import {render} from '@testing-library/react';
import {Provider} from 'react-redux';
import {createMemoryRouter, RouterProvider} from 'react-router-dom';
import {ToastContainer} from 'react-toastify';
import store from 'Store/configureStore';

interface WrapperForTestsParams {
    initialRouterEntries: string[];
}

export const setupTestsWrapper = ({initialRouterEntries}: WrapperForTestsParams) => {
    const memoryRouter = createMemoryRouter(router, {
        initialEntries: initialRouterEntries,
    });
    render(
        <Provider store={store}>
            <RouterProvider router={memoryRouter} />
            <ToastContainer autoClose={2000} position="top-right" theme="light" closeOnClick />
        </Provider>
    );

    return memoryRouter;
};
