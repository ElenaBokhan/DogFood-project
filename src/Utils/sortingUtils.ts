export const calculateOldPrice = (price: number, discount: number) => {
    return Math.floor(price - (price * discount) / 100);
};
