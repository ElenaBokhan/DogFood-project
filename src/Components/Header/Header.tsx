import cartIcon from 'assets/ic-cart.svg';
import profileIcon from 'assets/ic-profile.svg';
import Logo from 'assets/Logo.svg';
import logOut from 'assets/log-out.svg';
import addNewIcon from 'assets/add-new.svg';
import favouritesIcon from 'assets/profile-favorites.svg';
import HeaderContainer, {EContainerType} from 'Components/Common/Container/Container';
import styles from 'Components/Header/Header.module.css';
import {SearchForm} from 'Components/SearchForm/SearchForm';
import {useLocation, useNavigate} from 'react-router-dom';
import {IconButton} from 'Components/Common/IconButton/IconButton';
import {UseAppSelector} from 'Store/hooks';
import {selectCartCount} from 'Store/Slices/cart/CartSelectors';
import {selectFavouritesCount} from 'Store/Slices/favourites/FavouritesSelector';
import {useActions} from 'hooks/hooks';
import {ETestId} from 'Enum';
import { accessTokenSelector } from 'Store/Slices/auth/AuthSelectors';

export const Header = () => {
    const navigate = useNavigate();
    const {pathname} = useLocation();
    const {clearTokens} = useActions();
    const cartCount = UseAppSelector(selectCartCount);
    const favouriteCount = UseAppSelector(selectFavouritesCount);

    const accessToken = UseAppSelector(accessTokenSelector);

    const renderLogo = () => (
        <img data-testid={ETestId.HEADER_MAIN_LOGO} width={224} height={56} src={Logo} alt="logoMain" />
    );

    const redirect = (path: string) => () => navigate(path);

    const handleLogOut = () => {
        clearTokens();
        navigate('/signin');
    };

    const renderHeaderProfileIcons = () => {
        const headerIconConfig = [
            {
                name: favouritesIcon,
                onClick: redirect('/favourites'),
                babl: favouriteCount,
                testId: ETestId.HEADER_FAVOURITES_ICON,
            },
            {name: cartIcon, onClick: redirect('/cart'), babl: cartCount, testId: ETestId.HEADER_CART_ICON},
            {name: profileIcon, onClick: redirect('/profile'), testId: ETestId.HEADER_PROFILE_ICON},
            {name: addNewIcon, onClick: redirect('/addProduct'), testId: ETestId.HEADER_ADD_NEW_ICON},
            {name: logOut, onClick: handleLogOut, testId: ETestId.HEADER_LOG_OUT_ICON},
        ];

        return (
            <div className={styles.profilesIcons}>
                {headerIconConfig.map(({name, onClick, babl, testId}, index) => {
                    return (
                        <div key={index} className={styles.profilesIcon}>
                            {!!babl && <div className={styles.babl}>{babl}</div>}
                            <IconButton testId={testId} key={index} icon={name} onClick={onClick} alt="headerIcon" />
                        </div>
                    );
                })}
            </div>
        );
    };

    return (
        <HeaderContainer className={styles.header} type={EContainerType.HEADER}>
            {renderLogo()}
            {accessToken && pathname === '/' && <SearchForm />}
            {accessToken && renderHeaderProfileIcons()}
        </HeaderContainer>
    );
};

