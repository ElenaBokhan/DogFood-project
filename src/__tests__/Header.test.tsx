import '@testing-library/jest-dom';
import {setupTestsWrapper} from 'Utils/setupTestWrapper';
import {render, screen} from '@testing-library/react';
import * as Selectors from 'Store/Slices/auth/AuthSelectors';
import {ETestId} from 'Enum';
import store from 'Store/configureStore';
import {Provider} from 'react-redux';
import {Header} from 'Components/Header/Header';
import {BrowserRouter} from 'react-router-dom';

describe('Тесты для компонента Header', () => {
    beforeEach(() => {
        setupTestsWrapper({initialRouterEntries: ['/']});
    });

    describe('Проверяем доступность элементов под неавторизованным пользователем', () => {
        checkHeaderMainLogoIsAvailable();
        checkHeaderProfilesIconAvailable(false);
        checkHeaderSearchInputAvailable(false);
    });

    describe('Проверяем доступность ссылок под авторизованным пользователем', () => {
        beforeEach(() => {
            jest.spyOn(Selectors, 'accessTokenSelector').mockImplementation(() => 'token');
        });

        checkHeaderMainLogoIsAvailable();
        checkHeaderProfilesIconAvailable(true);
        checkHeaderSearchInputAvailable(true);
    });

    describe('Снапшот компонента Header', () => {
        test('Snapshot', () => {
            const {asFragment} = render(
                <Provider store={store}>
                    <BrowserRouter basename="/">
                        <Header />
                    </BrowserRouter>
                </Provider>
            );

            expect(asFragment()).toMatchSnapshot();
        });
    });
});

function checkHeaderMainLogoIsAvailable() {
    test('Проверяем доступность логотипа', () => {
        expect(screen.getByTestId(ETestId.HEADER_MAIN_LOGO)).toBeInTheDocument();
    });
}

function checkHeaderProfilesIconAvailable(enable: boolean) {
    if (enable) {
        test('Проверяем доступность иконок профиля', () => {
            expect(screen.getByTestId(ETestId.HEADER_FAVOURITES_ICON)).toBeInTheDocument();
            expect(screen.getByTestId(ETestId.HEADER_CART_ICON)).toBeInTheDocument();
            expect(screen.getByTestId(ETestId.HEADER_PROFILE_ICON)).toBeInTheDocument();
            expect(screen.getByTestId(ETestId.HEADER_LOG_OUT_ICON)).toBeInTheDocument();
        });
    } else {
        test('Проверяем не доступность иконок профиля', () => {
            expect(screen.queryByTestId(ETestId.HEADER_FAVOURITES_ICON)).not.toBeInTheDocument();
            expect(screen.queryByTestId(ETestId.HEADER_CART_ICON)).not.toBeInTheDocument();
            expect(screen.queryByTestId(ETestId.HEADER_PROFILE_ICON)).not.toBeInTheDocument();
            expect(screen.queryByTestId(ETestId.HEADER_LOG_OUT_ICON)).not.toBeInTheDocument();
        });
    }
}

function checkHeaderSearchInputAvailable(enable: boolean) {
    if (enable) {
        test('Проверяем доступность компонента поиска', () => {
            expect(screen.getByTestId(ETestId.HEADER_SEARCH_INPUT)).toBeInTheDocument();
        });
    } else {
        test('Проверяем не доступность компонента поиска', () => {
            expect(screen.queryByTestId(ETestId.HEADER_SEARCH_INPUT)).not.toBeInTheDocument();
        });
    }
}
