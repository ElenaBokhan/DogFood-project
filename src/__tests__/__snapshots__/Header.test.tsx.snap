// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Тесты для компонента Header Снапшот компонента Header Snapshot 1`] = `
<DocumentFragment>
  <header
    class="headerContainer"
  >
    <div
      class="container header"
    >
      <img
        alt="logoMain"
        data-testid="headerMainLogo"
        height="56"
        src="[object Object]"
        width="224"
      />
      <div
        class="searchForm"
      >
        <form>
          <input
            data-testid="headerSearchInput"
            placeholder="Поиск товаров"
            value=""
          />
        </form>
        <button
          class="button"
          type="button"
        >
          <img
            alt="searchButton"
            src="[object Object]"
          />
        </button>
      </div>
      <div
        class="profilesIcons"
      >
        <div
          class="profilesIcon"
        >
          <button
            class="button"
            type="button"
          >
            <img
              alt="headerIcon"
              data-testid="headerFavouritesIcon"
              src="[object Object]"
            />
          </button>
        </div>
        <div
          class="profilesIcon"
        >
          <button
            class="button"
            type="button"
          >
            <img
              alt="headerIcon"
              data-testid="headerCartIcon"
              src="[object Object]"
            />
          </button>
        </div>
        <div
          class="profilesIcon"
        >
          <button
            class="button"
            type="button"
          >
            <img
              alt="headerIcon"
              data-testid="headerProfileIcon"
              src="[object Object]"
            />
          </button>
        </div>
        <div
          class="profilesIcon"
        >
          <button
            class="button"
            type="button"
          >
            <img
              alt="headerIcon"
              data-testid="headerAddNewProduct"
              src="[object Object]"
            />
          </button>
        </div>
        <div
          class="profilesIcon"
        >
          <button
            class="button"
            type="button"
          >
            <img
              alt="headerIcon"
              data-testid="headerLogOutIcon"
              src="[object Object]"
            />
          </button>
        </div>
      </div>
    </div>
  </header>
</DocumentFragment>
`;
